lib/cn/cn_tree.c

```diff
@@ -451,47 +451,35 @@ tn_samp_update_finish(struct cn_tree_node *tn)
     s->ns_hclen = s->ns_kst.kst_halen;
     s->ns_pcap = min_t(uint16_t, UINT16_MAX, 100 * cn_ns_clen(s) / tn->tn_split_size);
 
-    tn->tn_samp.r_alen = 0;
-    tn->tn_samp.r_wlen = 0;
-
     if (cn_node_isleaf(tn)) {
-        tn->tn_samp.i_alen = 0;
+        tn->tn_samp.r_alen = 0;
         tn->tn_samp.l_alen = cn_ns_alen(s);
         tn->tn_samp.l_good = cn_ns_clen(s);
     } else {
-        tn->tn_samp.i_alen = cn_ns_alen(s);
+        tn->tn_samp.r_alen = cn_ns_alen(s);
         tn->tn_samp.l_alen = 0;
         tn->tn_samp.l_good = 0;
-
-        if (cn_node_isroot(tn)) {
-            tn->tn_samp.r_alen = cn_ns_alen(s);
-            tn->tn_samp.r_wlen = cn_ns_wlen(s);
-        }
     }
 }
 
 /* This function must be serialized with other cn_tree_samp_* functions. */
 static void
 cn_tree_samp_update_compact(struct cn_tree *tree, struct cn_tree_node *tn)
 {
-    bool                     need_finish = false;
-    struct cn_samp_stats     orig = tn->tn_samp;
     struct kvset_list_entry *le;
+    bool need_finish = false;
 
+    cn_samp_sub(&tree->ct_samp, &tn->tn_samp);
```

Comment:
This new sub/add interface makes things much clearer on what's going on here!!!

Reply:
Thanks!  Hopefully it'll also help prevent mishandling in the future, and make adding/removing fields to cn_samp_stats easier...